src/flag_gems/ops/conv2d.py

```diff
@@ -0,0 +1,384 @@
+            groups,
+            height_out,
+            width_out,
+            (c_out // groups),
+            dtype=input.dtype,
+            device=input.device,
+        )
+        groups = groups
+
+        grid = lambda meta: (
+            triton.cdiv(n_out, meta["BLOCK_M"]),
+            groups,
+            triton.cdiv((c_out // groups), meta["BLOCK_O"]),
+        )
+
+        conv2d_img2col[grid](
+            weight,
+            padded_input,
+            out,
+            padded_width_input,
+            padded_height_input,
+            height_kernel,
+            width_kernel,
+            c_kernel,
+            n_out,
+            stride_height,
+            stride_width,
+            padding_height,
+            padding_width,
+            height_out,
+            width_out,
+            groups,
+            (c_out // groups),
+        )
+
+        out = out.permute(0, 1, 4, 2, 3).contiguous()
+        out = torch.flatten(out, start_dim=1, end_dim=2)
+
+        weight_reshape = weight.reshape(-1, c_out)
+        ctx.save_for_backward(weight_reshape)
+        ctx.n_out = n_out
+        ctx.c_input = c_input
+        ctx.height_input = height_input
+        ctx.width_input = width_input
+        ctx.c_out = c_out
+        ctx.height_kernel = height_kernel
+        ctx.width_kernel = width_kernel
+        ctx.padded_height_input = padded_height_input
+        ctx.padded_width_input = padded_width_input
+        ctx.height_out = height_out
+        ctx.width_out = width_out
+        ctx.stride_height = stride_height
+        ctx.stride_width = stride_width
+
+        ctx.padding_height = padding_height
+        ctx.padding_width = padding_width
+        ctx.dtype = input.dtype
+        ctx.device = input.device
+        return out
+
+    @staticmethod
+    def backward(ctx, out_grad):
+        logging.debug("GEMS CONV2D VJP")
+        (weight_reshape,) = ctx.saved_tensors
+        c_input = ctx.c_input
+        padded_height_input = ctx.padded_height_input
+        padded_width_input = ctx.padded_width_input
+        n_out = ctx.n_out
+        dtype = ctx.dtype
+        device = ctx.device
+        padded_input = torch.zeros(
+            n_out,
+            c_input,
+            padded_height_input,
+            padded_width_input,
+            dtype=dtype,
+            device=device,
+        )
+
+        grid = lambda meta: (triton.cdiv(n_out, meta["BLOCK_M"]),)
+        height_kernel = ctx.height_kernel
+        width_kernel = ctx.width_kernel
+        stride_height = ctx.stride_height
+        stride_width = ctx.stride_width
+        height_out = ctx.height_out
+        width_out = ctx.width_out
+
+        stride_height = ctx.stride_height
+
+        padding_height = ctx.padding_height
+
+        c_out = ctx.c_out
+        out_grad = out_grad.permute(0, 2, 3, 1).reshape(-1, c_out)
+        input_col = mm(out_grad, weight_reshape.T)
+
+        # return dx,None,None,None,None,None,None
+        conv2d_col2img[grid](
```

Comment:
maybe we could fuse mm and col2img together as bwd kernel

Reply:
but not necessarily

Reply:
done